MLIR prototype

rengolin opened the pull request:
This is the initial prototype for the MLIR lowering from Verona code.

It has the main features:
1. Creates an MLIR Generator, that transforms the AST into MLIR nodes. For now it can only lower one file, `function.verona` and does so in an opaque MLIR format (no dialect). It does not have a dialect yet.
2. A prototype of a dialect, empty, but with the minimal structure of an out-of-tree dialect to follow up when we start having proper Verona nodes.
3. A thin layer between AST and MLIR, to help convert the semantics between the two. It's far from complete, but it's an idea on how we should interface these two completely different representations.
4. Adds a LIT + FileCheck testing structure with 3 tests (including `function.verona`).
5. Adds an LLVM build CI loop, where we cache the LLVM builds on Azure artefacts.
6. It uses the artefacts to test Verona without building LLVM from scratch. The artefact can also be used by developers on their machines via the same script the CI uses.

It has the main problems:
1. The Generator's public API is crude. It needs to account that we'll be using it to drive code through MLIR and LLVM, not always printing them to files.
2. Error handling is primitive. Uses the `Error` from `ast` for parsing the source files, but relies on throwing exceptions for the MLIR public API. This should be clearer once we have a better public API.
3. The Windows Debug image for LLVM is too large for our Verona builder and we run out of disk. All other tests pass, including Windows Release, so I'm confident this is just a space issue. We need to decide how we're going to fix that.
4. The code is precariously generating opaque types and operations and not checking for anything. There are various asserts spread around to help convey meaning, but until we have a robust interface between AST and MLIR, and between the opaque MLIR and the Verona dialect, we won't know for sure what's the best way to lower to MLIR.
5. Lowering to LLVM only works when the MLIR is in standard dialect, which the opaque syntax is clearly not. We need to lower to Standard dialect first, then to LLVM, and that will take a while. I just left the code there as boilerplate.

I'm expecting a long round of comments, changes, fixing the CI, etc. But I also expect this to not develop into a full blown MLIR engine before merging. This is a first approach to the problem, and it would be faster and easier to discuss and fix those problems upstream.

SeanTAllen commented on src/mlir/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+find_package(Threads REQUIRED)
+
+# We're not building LLVM, but downloading the build dir from an artefact
```
artefact is correct if you are British.

Theodus replied:
Ah, ok. Sorry

rengolin replied:
My syntax highlight conditions me to write British and the DevOps syntax checker conditions me to write American. Now, I have no idea wheter to salivate or pee in my pants when the bell rings.

rengolin commented on src/mlir/ast-utils.h:
```diff
@@ -0,0 +1,76 @@
+  // Find a sub-node of tag 'type'
+  ::ast::WeakAst findNode(::ast::WeakAst ast, NodeType::Int type);
+
+  // Get token value or return "unk{inc++}" if empty
```
I have no idea. :D

rengolin commented on src/mlir/README.md:
```diff
@@ -0,0 +1,23 @@
+        -DCMAKE_C_COMPILER=clang \
+        -DCMAKE_CXX_COMPILER=clang++ \
+        -DLLVM_ENABLE_LLD=ON \
+        -DLLVM_REQUIRES_EH=ON \
```
This is required by us. I may be building it wrong, but some external dependencies have exception handling and don't link well against LLVM's. I haven't dug deeply, tbh. Would be very happy if we don't need to do that.

rengolin replied:
Just to note that, currently, the MLIR generator uses EH, but that's a stop gap until we have decent error handling.

davidchisnall replied:
In the past, I used non-RTTI builds of LLVM with a front end that used RTTI and exceptions.  Everything was fine unless you subclassed an LLVM class.  I wonder if we have the same issue, not passing -fno-rtti to the builds of our MLIR things?

Someone accidentally missed off some anchors in LLVM's Instructions.h, which caused the vtable for a couple of LLVM classes to be emitted in every compilation unit, which was a problem, but I think that was fixed a few years back.

rengolin replied:
We do subclass a lot of MLIR stuff for the dialect. I really want to get rid of EH+RTTI, by probably using `llvm::ErrorOr`, but I haven't got around using it. The main reason is that I don't just want to treat errors as they happen, I want to compose a stack trace of errors and let the top caller handle it. This will be fundamental for partial compilation, pretty printing, IDE integration, etc.

rengolin commented on src/mlir/ast-utils.cc:
```diff
@@ -0,0 +1,127 @@
+    return *sub;
+  }
+
+  llvm::StringRef getTokenValue(::ast::WeakAst ast)
```
I'm glad you bring that up. :)

The ownership model here is completely "fell into place" kind of thing. The AST utils morphed into shape by developing the Generator and learning how to traverse the tree on the fly.

I'm hopeful the model will be simpler than what it is today. I don't like the weak pointer handling but had to use to work with the AST API. I think this deserves a more in depth discussion once we settle on an interface between the AST and MLIR.

rengolin commented on src/mlir/ast-utils.h:
```diff
@@ -0,0 +1,76 @@
+  // also creates an additional layer of security.
+  namespace NodeType
+  {
+    using Int = unsigned int;
```
I was playing with enum class and enums in namespaces, and this is so I can declare the argument of findNode via the NodeType's underlying type and not have conversion issues.

MLIR "handles" that by using the native type in argument declaration, I think that's horrible and error prone, so I wanted to name the type. `UInt` seems better, but still uinsatisfactory.

rengolin commented on src/mlir/ast-utils.h:
```diff
@@ -0,0 +1,76 @@
+  // constexpr enum type creation and let the rest be unsigned comparisons.
+  // The AST code needs to be flexible, so using the operator directly is more
+  // convenient. But we need to be very strict (with MLIR generation), so this
+  // also creates an additional layer of security.
```
Sorry, same word in Portuguese, and I still mix them today...

rengolin commented on src/mlir/ast-utils.h:
```diff
@@ -0,0 +1,76 @@
+  namespace NodeType
+  {
+    using Int = unsigned int;
+    enum NodeType : Int
```
`Kind` is probably better. But this is actually the AST type, not MLIR. I'll re-write that part to make more sense.

rengolin commented on src/mlir/dialect/VeronaDialect.cc:
```diff
@@ -0,0 +1,122 @@
+  if (parser.parseKeyword(&keyword))
+    return Type();
+
+  if (keyword.startswith("U"))
```
Sorry, lazyness is one of the three programmers virtues, right? Honestly, that was pro-activity of my part on `printType`. The types right don't mean much. I'll add `SNN` just to make more sense.

rengolin commented on src/mlir/dialect/VeronaDialect.cc:
```diff
@@ -0,0 +1,122 @@
+      Unsigned,
+      Signed
+    };
+    unsigned sign;
```
Because `std::get<1>(key)` returns unsigned, which is defined as `KeyTy` that way so `TypeStorage` can use it internally for a `DenseMap` key, given that `SignType` is a local type.

davidchisnall replied:
I don't think I understand.  Why can't `KeyType` be `std::tuple<size_t, SignType>`?

rengolin replied:
AFAICS, it's because that's used as a key to `DenseMap` and some of its own functionality isn't specialised for the local type. I haven't dug deep to make it work, but seeing the same pattern in other examples of MLIR, I just repeated the pattern and spent my time elsewhere.

We could perhaps revisit this issue once we have an actual type to implement.

rengolin commented on src/mlir/dialect/VeronaDialect.cc:
```diff
@@ -0,0 +1,122 @@
+{
+  struct IntegerTypeStorage : public ::mlir::TypeStorage
+  {
+    size_t width;
```
Sorry, `size_t` is my default (register sized) type, unless I need something else. I'll change to `uint8_t` so that we limit out integer sizes to 2^255, which I think it's good enough.

rengolin commented on src/mlir/dialect/VeronaDialect.cc:
```diff
@@ -0,0 +1,122 @@
+    unsigned sign;
+
+    // width, sign
+    using KeyTy = std::tuple<size_t, unsigned>;
```
Same answer.

rengolin commented on src/mlir/dialect/VeronaDialect.cc:
```diff
@@ -0,0 +1,122 @@
+    static IntegerTypeStorage*
+    construct(TypeStorageAllocator& allocator, const KeyTy& key)
+    {
+      return new (allocator.allocate<IntegerTypeStorage>())
```
This is only a stub, not a final implementation. That is the generic approach to type storage, whcih is applicable to all types. I'm leaving there more as documentation than actual implementation. None of the existing Verona dialect constructs will survive long enough, anyway.

rengolin commented:
@davidchisnall thanks for the comments. The questions that I marked "done" were addressed by the last commit.

Bear in mind that this is a prototype. The main intention here is to allow people to work on different parts of the compiler at the same time. This patch adds "AST Interface" and "Opaque MLIR lowering", and once the opaque MLIR is mature enough, we'll add the "Verona Dialect", and after that, "Lowering to Std dialects+LLVM".

Any of these new topics will start as a prototype, and slowly take shape. The good news is that, as we move away from the first patches, consensus is formed more easily and code quality improves significantly.

plietar commented:
On the PR itself, it seems constraints of the form `where X: T` are treated as aliases, which they aren't. The X should be a type variable, and `T` only represents its upper bound. We definitely want to use the type variable in the variables' types, and later teach the type checker about those constraints, so it can make appropriate subtyping decisions. For now I would just ignore the constraints completely (or just encode them in the MLIR metadata if that's easy enough), we'll use them once we actually implement the type checker.

I also have a few questions on the bigger picture of the MLIR pipeline, happy to discuss here or somewhere else:
- How will "non-executable" code be represented? By this I mean the Verona module structure (which might not match MLIR modules 1-1), class and interfaces definitions, the list of their fields, ... Will we be laying all method implementations at a single level in the MLIR module, with mangled names, and have a separate datastructure (maybe encoded in MLIR module metadata) that points back into that.

- Is the opaque MLIR a long term thing, or just a stop gap? It's not really clear to me what it actually looks like. At what stage does the SSA conversion happen, between AST and opaque MLIR, or is it an MLIR to MLIR lowering pass? The old prototype had a pre-type checking SSA IR (similar to what we'll have with the Verona dialect), and the conversion happens during the lowering from the AST.

- When running in "batch" mode (ie. source to binary, without reading or writing any intermediate results), would types have to go through a string representation? I would have expected MLIR to be able to carry structured data in memory, only going through a string representation when asked for a textual output, but that's not what the code seems to be doing. Same goes for other bits that we might want to store in MLIR metadata.

plietar commented:
Obviously I don't expect us to have all the answers at this point, I'm just curious at to what the plan might be.

davidchisnall commented on src/mlir/generator.h:
```diff
@@ -0,0 +1,102 @@
+
+namespace mlir::verona
+{
+  // MLIR Generator.
```
Please can you make doc comments something that a doc comment extractor can spot?  I prefer `/** */`, LLVM likes `///`.

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+  template<class T>
+  class ScopedTable
+  {
+    using MapTy = std::map<std::string, T>;
```
Why `map` not `unordered_map`?  Actually, `llvm::SmallDenseMap` probably makes more sense, since these are likely to be very small things.  It's probably worth using a new type for the key as well.  Eventually we're likely to want them to be interned strings, rather than doing string comparison on every lookup.

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    bool inScope(llvm::StringRef key)
+    {
+      auto frame = stack[currentScope];
```
```suggestion
auto frame = stack.back();
```
Pedantic: I prefer predicate methods to start with `is`, as in `isInScope`.

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+  {
+    using MapTy = std::map<std::string, T>;
+    std::vector<MapTy> stack;
+    int currentScope;
```
Is this ever any value other than `stack.size()`?  Do we need it?
```suggestion
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    int currentScope;
+
+  public:
+    ScopedTable() : currentScope(-1)
```
```suggestion
    ScopedTable()
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    ~ScopedTable()
+    {
+      popScope();
+      assert(stack.empty() && currentScope == -1);
```
```suggestion
      assert(stack.empty());
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    bool insert(llvm::StringRef key, T value)
+    {
+      auto& frame = stack[currentScope];
```
```suggestion
      auto& frame = stack.back();
```

rengolin commented:
Hi Paul,

Thanks for the great questions!

> On the PR itself, it seems constraints of the form `where X: T` are treated as aliases, which they aren't. The X should be a type variable, and `T` only represents its upper bound. We definitely want to use the type variable in the variables' types, and later teach the type checker about those constraints, so it can make appropriate subtyping decisions. For now I would just ignore the constraints completely (or just encode them in the MLIR metadata if that's easy enough), we'll use them once we actually implement the type checker.

That's right, I'm extremely oversimplifying for a practical reason: MLIR has strong type checking. If I just make the type of the argument to be "N" and later have an expression that expects it to be "U64 & imm", MLIR will barf. So I have used the alias approach to reduce all types to their minimum common sub-set. 

This won't work with any more complex example, obviously, but that's an entry point to discuss how we're going to implement it in MLIR. It's a lot easier to discuss implementation details when looking at a (broken) implementation already.

> * How will "non-executable" code be represented? By this I mean the Verona module structure (which might not match MLIR modules 1-1), class and interfaces definitions, the list of their fields, ... Will we be laying all method implementations at a single level in the MLIR module, with mangled names, and have a separate datastructure (maybe encoded in MLIR module metadata) that points back into that.

In discussions with Sylvan, we're considering a partial lowering to a flat representation, more or less how some C++ compilers lower to C first, then lower from C to IR. The big problem is that if we do that, then all work would have been on the AST and we won't get any goodness from MLIR. We need to find a middle ground where structures and functions are low enough that MLIR won't barf, but still retaining some cross-references and annotations that we can do Verona-specific passes before lowering to LLVM.

MLIR metadata and consistent name mangling will likely play an important part.

> * Is the opaque MLIR a long term thing, or just a stop gap? It's not really clear to me what it actually looks like. At what stage does the SSA conversion happen, between AST and opaque MLIR, or is it an MLIR to MLIR lowering pass? The old prototype had a pre-type checking SSA IR (similar to what we'll have with the Verona dialect), and the conversion happens during the lowering from the AST.

The current plan, which might very well change radically, is to keep the opaque MLIR for simplicity purposes, at least until we have a fixed AST/MLIR interface. It should be possible to translate from AST directly into Verona dialect (whatever that might end up being), but if we start now, we'll fix things that shouldn't be fixed, increasing the cost of improving the translation semantics and ending up with legacy code before moving out of the prototype stage.

I would hope that, in the production compiler, the AST would go directly into Verona dialect. But right now, we're not even sure this is the AST we'll have long term. So we're playing safe.

> * When running in "batch" mode (ie. source to binary, without reading or writing any intermediate results), would types have to go through a string representation? I would have expected MLIR to be able to carry structured data in memory, only going through a string representation when asked for a textual output, but that's not what the code seems to be doing. Same goes for other bits that we might want to store in MLIR metadata.

AFAICS, MLIR types are too simple and strict to do what we want on type inference. The opaque textual representation was an attempt to make it less strict, but turns out, the strings must be identical for the types to be considered compatible. So, the current string representaiton will have to change soon, and it's just there as a discussion point.

In Verona, objects can have more than one type and that's very alien to strongly typed intermediate representations, so we'll have basically two choices:

1. Do *all* type inference in the AST and lower strict types in MLIR, so that the verifier can be satisfied with the result.
2. Change MLIR to understand the concept of unions and intersections of types. This being a special case and a deep refactory, would probably not be accepted upstream, likely not without an existing concrete example (not using it).

In any case, we'll need to implement it without changing MLIR too much first, then propose changes to improve the code, not to enable it to work.

[Addendum]

I lied, there could be a third way, using `TypeStorage` classes to keep the metadata, and making all types from the same class, say `VeronaType`. This makes the verifier happy (as everything is guaranteed to have the same type), but then gets harder to query types (need to dig their internal data) and we'll have to implement type checking entirely from scratch.

We'd also have to work out how to serialise that, so that when we print the IR and read it back in, we get all information as before. This could be as attributes of the type (or the object), for example, but it's too soon to know how it will work.

I'm guessing in the end we'll end up doing a mix of all three choices.

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    T lookup(llvm::StringRef key)
+    {
+      for (int i = currentScope; i >= 0; i--)
+      {
+        auto& frame = stack[i];
```
```suggestion
      for (auto &frame : llvm::reverse(stack))
      {
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    bool update(llvm::StringRef key, T value)
+    {
+      auto& frame = stack[currentScope];
```
```suggestion
      auto& frame = stack.back();
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    void pushScope()
+    {
+      currentScope++;
```
```suggestion
```

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    void popScope()
+    {
+      currentScope--;
```
```suggestion
```

rengolin replied:
I need to decrease the counter to make sure the assert in the destructor passes.

rengolin replied:
Simplified the whole `ScopedTable` implementation.

davidchisnall commented on src/mlir/symbol.h:
```diff
@@ -0,0 +1,118 @@
+    void popScope()
+    {
+      currentScope--;
+      stack.resize(stack.size() - 1);
```
```suggestion
      stack.pop_back();
```

davidchisnall commented on src/mlir/verona-mlir.cc:
```diff
@@ -0,0 +1,228 @@
+namespace
+{
+  // Source file types, to choose how to parse
+  enum class Source
```
SourceKind?

rengolin replied:
ack

davidchisnall commented on src/mlir/verona-mlir.cc:
```diff
@@ -0,0 +1,228 @@
+  enum class Source
+  {
+    NONE,
+    VERONA,
```
Why are these capitalised?  Other enums are not.

rengolin replied:
no reason, will change

davidchisnall commented on src/mlir/verona-mlir.cc:
```diff
@@ -0,0 +1,228 @@
+  // Parse cmd-line and set defaults
+  Opt parse(int argc, char** argv)
+  {
+    CLI::App app{"Verona MLIR"};
```
Is this command-line arg thing significantly better than the LLVM one?  Since we're already parsing LLVM args, I'd prefer to not have two command-line arg parsers...

rengolin replied:
I agree. I'm using it because of how the AST parser used to work (had to pass opt). Now it passes the internal results only, so it's not necessary anymore.

I think this could be a following commit, in order to get the big one out of the way.

rengolin commented on src/mlir/dialect/VeronaOps.td:
````diff
@@ -0,0 +1,37 @@
+        ```
+    }];
+
+    let arguments = (ins Verona_Type:$input);
````
Correct. This is a useless op, but is a placeholder for other Verona ops. The restrictions are purely arbitrary.

rengolin replied:
Done.

rengolin commented on src/mlir/generator.cc:
```diff
@@ -0,0 +1,350 @@
+    if (mlir::failed(mlir::verify(*module)))
+    {
+      module->dump();
+      throw std::runtime_error("MLIR verification failed from Verona file");
```
We don't. This is a stop-gap until we have a better scheme. I'm planning to look at it right after this PR, but didn't want to add another long discussion to this already big PR.

FYI, I'm trying to come up with an object that is (Value | Error) and a set of helpers to handle that gracefully and stack as it's returned to a caller that actually handles it. If that doesn't work, having class-wide stack tracing (also stackable into other class-wide error handlers) is my second choice. Passing around error objects (like the AST does) is my third choice and returning `mlir::success()` is my last one.

davidchisnall replied:
`llvm::ErrorOr`?

rengolin replied:
Yup, top of the list.

rengolin commented on src/mlir/generator.cc:
```diff
@@ -0,0 +1,350 @@
+      throw std::runtime_error("No input filename provided");
+
+    // Read an MLIR file
+    llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> srcOrErr =
```
copy&paste. :) will fix.

rengolin commented on src/mlir/generator.cc:
```diff
@@ -0,0 +1,350 @@
+
+    // Setup source manager and parse
+    llvm::SourceMgr sourceMgr;
+    sourceMgr.AddNewSourceBuffer(std::move(*srcOrErr), llvm::SMLoc());
```
Good point. I came from a mindset where "any error crashes the compiler" to "trying to use EH to handle some errors" and that one missed the mark. Will fix.

rengolin commented on src/mlir/generator.cc:
```diff
@@ -0,0 +1,350 @@
+    if (filename.empty())
+      throw std::runtime_error("No output filename provided");
+
+    // The lowering "pass manager"
```
Not only that, but it must be in a dialect that has implemented full conversion into LLVM dialect. This code is only placeholder, it is not being called and is guaranteed to break. I can remove, f you think it's premature.

davidchisnall replied:
No, it's good to have it here, it would just be nice to have a comment explaining how it fits into the big picture.

rengolin replied:
added a comment

rengolin commented:
Thanks David!

Fixed some of the comments, and marked them as "resolved".

The 4 tasks to do are:
1. Remove `emitLLVM` for now, if you think it's premature to leave it there, unused.
2. Change error handling, perhaps to something like `llvm::ErrorOr` or a new similar class.
3. Refactor the symbol table, assuming we'll never need to search on specific contexts other than descending order from the top.
4. Change the cmdline argument handling to LLVM, now that we don't need it anymore for AST.

The first one I can do now, if you think it's necessary. The other three I'd rather do it on separate PRs.

davidchisnall commented:
Sounds good.  I agree it's better to land this and then improve it.

rengolin commented:
Ok, it passes all the tests, but only when I trigger by hand, as the [GitHub permission problem](https://developercommunity.visualstudio.com/content/problem/1087365/universal-package-download-does-not-exist-or-you-d.html) is still there. The "microsoft.verona" build doesn't complete because I moved the builds in the branch `devops`, which this PR depends on, so that build won't exist anymore.

I'll continue to investigate, given the VS folks are not responding. Hopefully it will be some silly permission problem and we'll be able to continue with our work.

Please don't merge until we can trigger a successful build straight from GitHub.

rengolin commented:
Also depends on #162 merging first, so we can remove "microsoft.verona" form the required list and have a green status.

rengolin commented:
I have rebased and squashed the commits into one big patch. The commit message was updated to reflect all the work done including the review response. Hopefully, further commits will be a lot smaller, needing a lot less disruption to the build system.

Right now, the build works, the CI passes and the Nightly run only fails because of the gcc issues. The LLVM build works and puts the tarballs in the right place, the CI and Nightly runs pull them correctly and the build works as expected. There are no more holders in this PR, we can merge it as soon as the tests clear up.

Thank you for the help and patience in making this huge merge work. Now I'll be able to go back doing normal work. :)

Btw, there is an issue with download speeds in the feed. Usually, it takes 10 seconds or less, sometimes it can take up to 20 minutes. AFAICS, it affects Windows and Mac, which tells me it's not PowerShell, wget or curl, neither is the image itself. I think it's the blob storage. I got premium disks, but I may be doing something silly on the configuration. The storage is on the same zone as the builders.

@mjp41 @sylvanc @davidchisnall @plietar @Theodus @SeanTAllen
